internal/validation/apiextensions/v1/composition/handler.go

```diff
@@ -96,3 +118,132 @@ func validationResponseFromStatus(allowed bool, status metav1.Status) admission.
+	for _, err := range errs {
+		if err == nil {
+			continue
+		}
+		// If any of the errors is not a NotFound error, error out
+		if !apierrors.IsNotFound(err) {
+			return false, warns, xperrors.Errorf("there were some errors while getting the needed CRDs: %v", errs)
+		}
+		// If any of the needed CRDs is not found, error out if strict mode is enabled
+		switch validationMode {
+		case v1.CompositionValidationModeStrict:
+			return false, warns, xperrors.Wrap(err, "cannot get needed CRDs and strict mode is enabled")
+		case v1.CompositionValidationModeLoose:
+			// Given that some requirement is missing, and we are in loose mode, skip validation
+			shouldSkip = true
+			warns = append(warns, fmt.Sprintf("cannot get needed CRDs and loose mode is enabled: %v", err))
+			continue
+		}
+	}
+	return shouldSkip, warns, nil
+}
+
+func (h *handler) getNeededCRDs(ctx context.Context, comp *v1.Composition) (map[schema.GroupVersionKind]apiextensions.CustomResourceDefinition, []error) {
+	var resultErrs []error
+	neededCrds := make(map[schema.GroupVersionKind]apiextensions.CustomResourceDefinition)
+
+	// Get schema for the Composite Resource Definition defined by comp.Spec.CompositeTypeRef
+	compositeResGVK := schema.FromAPIVersionAndKind(comp.Spec.CompositeTypeRef.APIVersion,
+		comp.Spec.CompositeTypeRef.Kind)
+
+	compositeCRD, err := h.getCRDForGVK(ctx, &compositeResGVK)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return nil, []error{err}
+		}
+		resultErrs = append(resultErrs, err)
+	}
+	if compositeCRD != nil {
+		neededCrds[compositeResGVK] = *compositeCRD
+	}
+
+	// Get schema for all Managed Resource Definitions defined by comp.Spec.Resources
+	for _, res := range comp.Spec.Resources {
+		cd, err := res.GetBaseObject()
+		if err != nil {
+			return nil, []error{err}
+		}
+		gvk := cd.GetObjectKind().GroupVersionKind()
+		if _, ok := neededCrds[gvk]; ok {
+			continue
+		}
+		crd, err := h.getCRDForGVK(ctx, &gvk)
+		switch {
+		case apierrors.IsNotFound(err):
+			resultErrs = append(resultErrs, err)
+		case err != nil:
+			return nil, []error{err}
+		case compositeCRD != nil:
+			neededCrds[gvk] = *crd
+		}
+	}
+
+	return neededCrds, resultErrs
+}
+
+// getCRDForGVK returns the validation schema for the given GVK, by looking up the CRD by group and kind using
+// the provided client.
+func (h *handler) getCRDForGVK(ctx context.Context, gvk *schema.GroupVersionKind) (*apiextensions.CustomResourceDefinition, error) {
+	crds := extv1.CustomResourceDefinitionList{}
+	if err := h.reader.List(ctx, &crds,
+		client.MatchingFields{"spec.group": gvk.Group},
+		client.MatchingFields{"spec.names.kind": gvk.Kind}); err != nil {
+		return nil, err
+	}
+	if len(crds.Items) != 1 {
+		return nil, apierrors.NewNotFound(schema.GroupResource{Group: "apiextensions.k8s.io", Resource: "CustomResourceDefinition"}, fmt.Sprintf("%s.%s", gvk.Kind, gvk.Group))
```

Comment:
It should not be possible, as those values are enforced to be unique by the api-server, nonetheless makes sense to return different errors, maybe an internal server error in case of more than 1?

Reply:
`internal server error` sounds good to me.